src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageApplicationModel.cs

```diff
@@ -109,7 +109,15 @@ public PageApplicationModel(PageApplicationModel other)
         public TypeInfo PageType { get; set; }
 
         /// <summary>
-        /// Gets or sets the <see cref="TypeInfo"/> of the Razor page model.
+        /// Gets or sets the declared model <see cref="TypeInfo"/> of the model for the page.
+        /// Typically this <see cref="TypeInfo"/> will be the type of the Model property in the razor page.
+        /// </summary>
+        public TypeInfo DeclaredModelType { get; set; }
```

Comment:
So here's where it gets real. In 2.0 you could have written code to modify the model property, which would have worked fine.

When you run against 2.1, you would modify the model property but not the declaredmodeltype property, which could result is broken things.... 

We should do some thinking about how to mitigate this....

One possible (ugly) strategy would be to make `DeclaredModelType` null by default, and explain that if you want `DeclaredModelType` to be a different type than `ModelType` you should set both of them.

Any other ideas?

Reply:
If your page was typed you would have gotten the same casting exceptions I got all over the place

Reply:
Unless you decided to re-implement every point this thing touches. My point being, possible yes, but not likely. I think we've done way more severe things in other areas

Reply:
Updated

Reply:
> One possible (ugly) strategy would be to make DeclaredModelType null by default, and explain that if you want DeclaredModelType to be a different type than ModelType you should set both of them.

I didn't read the entire comment. This option sounds fine to me. I'm going to give it a spin locally and see what breaks. It's trivially more work for the implementor

Reply:
I've file a bug to track this, I would like us to move forward with the current design and look at this on preview2 https://github.com/aspnet/Mvc/issues/7297

Reply:
I think this is something simple we can just discuss and make a decision about. I don't think a follow issue is necessary.